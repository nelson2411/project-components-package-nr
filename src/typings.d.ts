/*
Here we define the types for our "styles.module.css" file.
*/

declare module '*.css' {
  const content: { [className: string]: string };
  export default content;
}

declare module '*.jpg' {
  const value: any;
  export default value;
}
