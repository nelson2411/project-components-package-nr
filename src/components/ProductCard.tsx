import React, { createContext, JSX } from "react"
import styles from "../styles/styles.module.css"
import { useProduct } from "../hooks/useProduct"
import {
  ProductContextProps,
  Product,
  onChangeArgs,
  InitialValues,
  ProductCardhandlers,
} from "../interfaces/ProductInterfaces"

export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext

export interface Props {
  product: Product
  children: (args: ProductCardhandlers) => JSX.Element // Children is a function that receives the product handlers
  className?: string // Optional className for custom styling. We defined it as optional to allow flexibility in styling the component.
  style?: React.CSSProperties // Optional style prop for inline styles
  onChange?: (args: onChangeArgs) => void // Callback function to handle changes in product count
  value?: number // Optional value prop to set the initial count of the product
  initialValues?: InitialValues // Optional initial values for the product, including count and maxCount
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const { counter, increaseBy, maxCount, isMaxCountReached, reset } =
    useProduct({
      onChange,
      product, // Passing the product to the useProduct hook
      value, // Passing the initial value to the useProduct hook
      initialValues, // Passing the initial values to the useProduct hook
    })

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product, // Providing the product to the context
        maxCount,
      }}
    >
      <div
        className={`${styles.productCard} ${className}`}
        style={style} // Applying the style prop to the root div
      >
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
          reset, // Providing the reset function to the children
        })}
      </div>
    </Provider>
  )
}
