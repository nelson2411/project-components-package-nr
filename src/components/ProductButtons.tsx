import React, { useContext } from "react"
import styles from "../styles/styles.module.css"
import { ProductContext } from "./ProductCard"

export interface Props {
  className?: string
  style?: React.CSSProperties // Optional style prop for inline styles
}

export const ProductButtons = ({ className, style }: Props) => {
  // ToDo: maxCount
  // Use the ProductContext to access the product state
  // This context should provide the current count and a function to increase the count

  const { counter, increaseBy, maxCount } = useContext(ProductContext)

  console.log("maxCount", maxCount)

  // ToDo: isMaxCountReached using a useCallback

  const isMaxCountReached = React.useCallback(
    () => !!maxCount && counter === maxCount,
    [counter, maxCount]
  )

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        className={`${styles.buttonAdd} ${
          isMaxCountReached() ? styles.disabled : ""
        }`}
        onClick={() => increaseBy(1)}
      >
        +
      </button>
    </div>
  )
}
