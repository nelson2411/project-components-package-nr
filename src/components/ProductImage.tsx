import React, { useContext } from "react"
import styles from "../styles/styles.module.css"
import { ProductContext } from "./ProductCard"
import noImage from "../assets/no-image.jpg"

export interface Props {
  img?: string // Optional img prop
  className?: string // Optional className for custom styling
  style?: React.CSSProperties // Optional style prop for inline styles
}

export const ProductImage = ({ img, className, style }: Props) => {
  const { product } = useContext(ProductContext)
  let imgToShow: string

  if (img) {
    imgToShow = img
  } else if (product.img) {
    imgToShow = product.img
  } else {
    imgToShow = noImage
  }

  return (
    <img
      src={imgToShow}
      alt="Product"
      className={`${styles.productImg} ${className}`}
      style={style} // Applying the style prop to the img element
    />
  )
}
