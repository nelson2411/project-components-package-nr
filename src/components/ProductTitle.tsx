import React, { useContext } from "react"
import styles from "../styles/styles.module.css"
import { ProductContext } from "./ProductCard"

// I need to create a Props interface that expects a title prop and a className optional prop.
// this interface will be used to type the props of the ProductTitle component.

export interface Props {
  title?: string // Optional title prop
  className?: string // Optional className for custom styling
  style?: React.CSSProperties // Optional style prop for inline styles
}

export const ProductTitle = ({ title, className, style }: Props) => {
  const { product } = useContext(ProductContext)

  return (
    <span className={`${styles.productTitle} ${className} `} style={style}>
      {/* If title is provided, use it; otherwise, use product.title */}
      {title ? title : product.title}
    </span>
  )
}
