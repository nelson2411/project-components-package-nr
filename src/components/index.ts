import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductTitle } from './ProductTitle';
import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductCardHOCProps } from '../interfaces/ProductInterfaces';

export { ProductButtons } from './ProductButtons';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
  Image: ProductImage,
  Title: ProductTitle,
  Buttons: ProductButtons,
});

export default ProductCard;
