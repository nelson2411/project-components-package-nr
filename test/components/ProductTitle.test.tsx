import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../data/products';

describe('ProductTitle', () => {
  test('renders without crashing', () => {
    const wrapper = renderer.create(
      <ProductTitle title="Test Product" className="custom-class" />
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('renders the component with the product name', () => {
    // here we will test the ProductCard with ProductTitle
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
