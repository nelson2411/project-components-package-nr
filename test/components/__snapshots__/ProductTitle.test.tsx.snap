// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle renders the component with the product name 1`] = `null`;

exports[`ProductTitle renders without crashing 1`] = `null`;
